### ISSUE TITLE
Recursive ADTs's derived Gen instances tends to OOM

### ISSUE THREAD

**Issue by tabdulradi** [zio/zio#3821]
Hello, 

Using DeriveGen with a recursive data structure like a Tree produces a Gen that takes reaaaaly long time to run before I kill it or I get OOM. Here is an example:
```scala
package com.abdulradi.ziotestrecursive

import zio.test._
import zio.test.magnolia._

sealed trait Tree
case object Leaf extends Tree
final case class Node(element: Int, children: List[Tree]) extends Tree

object Recursive extends DefaultRunnableSpec {
  private val derived = DeriveGen[Tree]
  def spec = suite("Recursive Spec")(
    testM("simple") { 
      println("starts")
      check(derived) { a => 
        println(a)
        assertCompletes
      }
    }
  )
}
```

Running this, I usually see one or two instances being created, then silence for long time and eventually it OOMs.

Even if I tried to limit branches, by using a binary tree instead, the behaviour becomes slightly better: sometimes I see 5 or 6 instances get printed. But after that it gets stuck and OOMs as usual.

The only branching factor that works is 1, so a hack that I am currently using is:
```scala
implicit val atMostOneNode: DeriveGen[List[Tree]] = 
    DeriveGen.instance(for {
      leafs <- Gen.listOf(DeriveGen[Leaf.type])
      nodes <- Gen.listOfBounded(0, 1)(DeriveGen[Node])
    } yield leafs ++ nodes)
```

Is it worth it to bake a max depth into the generated code to limit nested ADT from going crazy?

**Comment by adamgfraser** [comment]
@tabdulradi Yes that is definitely an issue. As you say it is an issue for recursive data types with a branching factor greater than one in general that we normally address with some concept of sized generators. The challenge is how to do this fully generically, because if we just have a `sealed trait` we don't know what the "base case" is that we we should be falling back to if we don't want to recurse. I've spent some time thinking about how to address this but will try to come back to it in the near future.
